mantidproject/mantid
# Adjust confidence band interval calculation for calculated fit curve

**Description of work.**
This PR corrects the method used to determine confidence bands on the calculated fit curve. Previously it was wrongly using the full chi square rather than the reduced chi square - explaining large errors bars seen on fit data. This PR also adds the student-t disturbution parameter to the bands, which was omitted before (as for 1 sigma confidence its approx equal to 1 for large number of degree of freedom). The underlying theory can be found at https://www.astro.rug.nl/software/kapteyn/kmpfittutorial.html

I had to update lots of reference files used in the systemtests in this PR due to the changes in error calculations. For each test I compared the previous and new reference calculations and observed that the only difference in each case was the error bars, confirming that everything else had remained the same.



**To test:**
Test workbench fitting and the linked issue to make sure the error bars are releastic. 
The following script compares mantid, lmfit and kmpfit, which have to be pip installed. This should show that the error bars from mantid and kmpfit are now identical.

```
from mantid.simpleapi import *
import numpy
from numpy.random import normal
from kapteyn import kmpfit
import matplotlib.pyplot as plt
from numpy import exp, loadtxt, pi, sqrt
from lmfit import Model
import scipy
from scipy.stats import t
from matplotlib.patches import Polygon

def confidence_band(x, dfdp, confprob, fitobj, f, abswei=False):
   alpha = 1 - confprob/100.0
   prb = 1.0 - alpha/2
   tval = t.ppf(prb, fitobj.dof)
   C = fitobj.covar
   n = len(fitobj.params)              # Number of parameters from covariance matrix
   p = fitobj.params
   N = len(x)
   if abswei:
      covscale = 1.0
   else:
      covscale = fitobj.rchi2_min
   df2 = numpy.zeros(N)
   for j in range(n):
      for k in range(n):
         df2 += dfdp[j]*dfdp[k]*C[j,k]
   df = numpy.sqrt(fitobj.rchi2_min*df2)
   y = f(p, x)
   delta = tval * df   
   upperband = y + delta
   lowerband = y - delta 
   return y, upperband, lowerband

# the models
def model(p, x):
   # Model: Y = a + b*x + c*x*x
   a,b,c  = p
   return a + b*x + c*x*x;

def quadratic(x, a0, a1, a2):
    return a0 + a1*x + a2*x*x
    
def residuals(p, data):  # Residuals function needed by kmpfit
   x, y = data           # Data arrays is a tuple given by programmer
   return (y-model(p,x))


function = FunctionFactory.Instance().createInitialized("name=Polynomial, n=2")


# CREATE THE DATA
N = 100
a0 = 8/9.; b0 = -2/9.; c0 = 1/9.
x = numpy.linspace(-3, 8.0, N)
y = model((a0,b0, c0), x) + normal(0.0, 0.5, N)  # Mean 0, sigma 1
residual_func = residuals
ws = CreateWorkspace(DataX=x,DataY=y, EnableLogging=False)


# KMP_FIT
beta0 = [1,-0.2,0.1]
fitobj = kmpfit.Fitter(residuals=residuals, data=(x, y))
fitobj.fit(params0=beta0)

dfdp = [1, x, x*x]
confprob = 68.268949213
ydummy, upperband, lowerband = confidence_band(x, dfdp, confprob, fitobj, model)
verts = list(zip(x, lowerband)) + list(zip(x[::-1], upperband[::-1]))
poly = Polygon(verts, closed=True, fc='b', ec='b', alpha=0.3,
               label="CI (1 sigma%)")

# CREATE THE FIGURE
fig, ax = plt.subplots(1, 4, subplot_kw={'projection':'mantid'}, figsize=(10,7))
plt.rc('font', size=9)
plt.rc('legend', fontsize=7)
ax[0].plot(x, y, 'bo')
ax[0].plot(x, model(fitobj.params,x), 'r', ls='--', lw=2, label="kmpfit")
ax[0].add_patch(poly)
ax[0].set_title("KmpFit")

print("\nFit status kmpfit:")
print("====================")
print("Best-fit parameters:        ", fitobj.params)
print("Error assuming red.chi^2=1: ", fitobj.stderr)
print("Chi^2 min:                  ", fitobj.chi2_min)
print("Reduced Chi^2:              ", fitobj.rchi2_min)

gmodel = Model(quadratic)
result = gmodel.fit(y, x=x, a0=0, a1=1, a2=1, scale_covar=False)
print(result.covar)

print("---------LMFit Results---------")
print(result.fit_report())
dely = result.eval_uncertainty(x=x, sigma=1)
ax[1].plot(x, y, 'bo')
ax[1].plot(x, result.best_fit)
ax[1].fill_between(x, result.best_fit-dely,
                 result.best_fit+dely, color='#888888')
ax[1].set_title("LMFit")


# MANTID FIT
ax[2].plot(ws, 'bo')
fit = Fit(function, ws, Output='out', EnableLogging=False, CalcErrors=True)
out_ws = fit.OutputWorkspace
paramTable = fit.OutputParameters  
# plotting
ax[2].errorbar(out_ws, specNum=2)
ax[2].set_title("Mantid")


print("---------Mantid Results---------")
print("The fit was: {}".format(fit.OutputStatus))
print("chi-squared of fit is: {:.5f}".format(fit.OutputChi2overDoF))
print("Fitted a0 value is: {:.5f} +- {:.5f}".format(paramTable.column(1)[0],paramTable.column(2)[0]))
print("Fitted a1 value is: {:.5f} +- {:.5f}".format(paramTable.column(1)[1],paramTable.column(2)[1]))

y = out_ws.readE(1)
kmp_error = (upperband - lowerband)/2
ax[3].plot(x, y-dely, 'r-', label="CI_MA - CI_LMF")
ax[3].plot(x, y-kmp_error, 'b--', label="CI_MA- CI_KMPF")
ax[3].set_title("CI difference")
ax[3].legend()

plt.show()

```


Fixes #28343 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
